prov/mrail/src/mrail_init.c

```diff
@@ -157,110 +154,132 @@ int mrail_get_core_info(uint32_t version, const char *node, const char *service,
-		FI_WARN(&mrail_prov, FI_LOG_CORE,
-			"Max mrail_num_info reached\n");
-		assert(0);
-		return -FI_ENODATA;
-	}
+	// Clears the cache, mrail_info_vec only caches the latest getinfo
+	for (i = 0; mrail_info_vec[i]; i++)
+		fi_freeinfo(mrail_info_vec[i]);
+	memset(mrail_info_vec, 0, sizeof(mrail_info_vec));
+	mrail_num_info = 0;
 
-	ret = mrail_get_core_info(version, node, service, flags, hints, info);
+	ret = mrail_get_core_info(version, node, service, flags, hints, &core_info);
 	if (ret)
 		return ret;
 
-	for (fi = *info, num_rails = 0; fi; fi = fi->next, ++num_rails)
+	for (num_rails = 0; mrail_addr_strv[num_rails]; ++num_rails)
 		;
 
 	mr_key_size = num_rails * sizeof(struct mrail_addr_key);
 
-	fi = fi_dupinfo(*info);
-	if (!fi) {
-		ret = -FI_ENOMEM;
-		goto err1;
-	}
-
-	free(fi->fabric_attr->name);
-	free(fi->domain_attr->name);
+	while (core_info) {
+		fi = core_info;
+		free(fi->fabric_attr->name);
+		free(fi->domain_attr->name);
 
-	fi->fabric_attr->name = NULL;
-	fi->domain_attr->name = NULL;
+		fi->fabric_attr->name = NULL;
+		fi->domain_attr->name = NULL;
 
-	fi->fabric_attr->name = strdup(mrail_info.fabric_attr->name);
-	if (!fi->fabric_attr->name) {
-		ret = -FI_ENOMEM;
-		goto err2;
+		fi->fabric_attr->name = strdup(mrail_info.fabric_attr->name);
+		if (!fi->fabric_attr->name) {
+			ret = -FI_ENOMEM;
+			goto err;
+		}
+		fi->domain_attr->name = strdup(mrail_info.domain_attr->name);
+		if (!fi->domain_attr->name) {
+			ret = -FI_ENOMEM;
+			goto err;
+		}
+		fi->ep_attr->protocol		= mrail_info.ep_attr->protocol;
+		fi->ep_attr->protocol_version	= mrail_info.ep_attr->protocol_version;
+		fi->fabric_attr->prov_version	= FI_VERSION(MRAIL_MAJOR_VERSION,
+							     MRAIL_MINOR_VERSION);
+		fi->domain_attr->mr_key_size	= mr_key_size;
+		fi->domain_attr->mr_mode        |= FI_MR_RAW;
+
+		/* Account for one iovec buffer used for mrail header */
+		assert(fi->tx_attr->iov_limit);
+		fi->tx_attr->iov_limit--;
+
+		/* Claiming messages larger than FI_OPT_BUFFERED_LIMIT would consume
+		 * a scatter/gather entry for mrail_hdr
+		 */
+		fi->rx_attr->iov_limit--;
+
+		if (fi->tx_attr->inject_size < sizeof(struct mrail_hdr))
+			fi->tx_attr->inject_size = 0;
+		else
+			fi->tx_attr->inject_size -= sizeof(struct mrail_hdr);
+
+		if (hints && hints->tx_attr && (hints->tx_attr->op_flags & FI_COMPLETION))
+			fi->tx_attr->op_flags |= FI_COMPLETION;
+
+		// TODO set src_addr to FI_ADDR_STRC address
+
+		for (i = 0; i < mrail_num_info; i++) {
+			if (!strcmp(core_info->fabric_attr->prov_name,
+			    mrail_info_vec[i]->fabric_attr->prov_name)) {
+				fi = mrail_info_vec[i];
+				while (fi->next)
+					fi = fi->next;
+				fi->next = core_info;
+				fi = fi->next;
+				core_info = core_info->next;
+				fi->next = NULL;
+				break;
+			}
+		}
+		if (!mrail_info_vec[i]) {
+			assert(mrail_num_info < MRAIL_MAX_INFO);
+			mrail_info_vec[i] = core_info;
+			core_info = core_info->next;
+			mrail_info_vec[i]->next = NULL;
+			mrail_num_info++;
+		}
```

Comment:
What is being done here is the assembly of the array of lists. The for loop checks if the provider is stored in the array, if the list exists, then append the new info to the end of that list. If it's not, create a new list entry in the array

The way the cache looks is:

```
mrail_info_vec:
	[0]: info_rail0 -> info_rail1 -> ... // sockets
	[1]: info_rail0 -> info_rail1 -> ... // tcp
        ...
```

At line 253, the info being sent back to the caller is being assembled. This info will have the format of:

```
mrail_info_hdr_prov1 -> mrail_info_hdr_prov2 -> ...
```

Instead of using mrail_info_vec from storing multiple getinfo calls, I changed it to store one getinfo call. For the case where the info given does not match anything in the cache on fabric open, I added functionality to cache a new getinfo call. The idea was that in the majority of cases, only the last getinfo would be used, and to cache only that entry. If you think that caching every getinfo call is necessary, I can change the structure to accommodate.

Reply:
Thank you for the clarification. Please add the above as a comment. I wonder if this can be simplified by building mrail_info_vec when we fetch core fi_infos for a rail.

I think we're fine with maintaining the cache for most recent fi_info. We can add support for caching multiple later if needed. We should at least print a warning when we're overwriting the cache.

Btw, preparing the hdr info (modifying the attributes) should be done after dupinfo. Are you not seeing any issue without this?

Reply:
I'm not exactly sure what you mean by modifying the attributes after the dupinfo. I was intending to modify the attributes for every info that is inside the cache, not just a hdr info, and I haven't seen any issues so far.

Reply:
The modifications are needed only for the "header info" that's returned to the app. It does things like adjusting the max iov size supported. We might end up requesting wrong attributes from the underlying provider if we modify the cache. e.g. fi_endpoint call directly passes the cached fi_info as hints to underlying provider. This is the reason the present code duplicates the fi_info before modifying.

Reply:
Oh okay, I see what you did there. I was wondering why you didn't apply the attributes to all the info structs and were maintaining a header info. I will see what I can do here. Moving the creation of the info cache into get core info would make more sense in this case (I originally had it do this before I changed it to the current implementation). Then we won't cache any of the modified attributes, only the caller will get the modified structs.

Reply:
Got it. If creating the cache when getting core info ends up being simpler, we would want to do that.